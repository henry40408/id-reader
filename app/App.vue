<template>
  <main class="container my-4 mx-auto font-sans">
    <RouterView />
  </main>
</template>
