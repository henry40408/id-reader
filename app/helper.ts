import pangu from 'pangu';

export function spacing(t: string): string {
  return pangu.spacingText(t);
}
