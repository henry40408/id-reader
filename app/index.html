<!doctype html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>id-reader</title>
  </head>
  <body>
    <div id="app"></div>
    <script type="module">
      import { DefaultApolloClient } from '@vue/apollo-composable';
      import { createApp, h, provide } from 'vue';
      import { createRouter, createWebHistory } from 'vue-router';
      import { apolloClient } from './apollo';

      import App from './App.vue';

      const routes = [
        { path: '/', component: () => import('./views/HomeView.vue') },
        { path: '/categories', component: () => import('./views/Categories.vue') },
      ];
      const router = createRouter({
        history: createWebHistory(),
        routes,
      });

      const app = createApp({
        setup() {
          provide(DefaultApolloClient, apolloClient);
        },
        render: () => h(App),
      });
      app.use(router).mount('#app');
    </script>
  </body>
</html>
