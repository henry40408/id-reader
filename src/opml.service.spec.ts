import fs from 'node:fs';
import path from 'node:path';
import { Test, TestingModule } from '@nestjs/testing';
import { getRepositoryToken } from '@nestjs/typeorm';
import { Repository } from 'typeorm';
import { AppModule } from './app.module';
import { CategoryEntity } from './entities/category.entity';
import { UserEntity } from './entities/user.entity';
import { DEFAULT_CATEGORY_NAME, OpmlService } from './opml.service';

describe('OPML service', () => {
  let moduleRef: TestingModule;
  let service: OpmlService;
  let user: UserEntity;

  beforeEach(async () => {
    moduleRef = await Test.createTestingModule({
      imports: [AppModule],
    }).compile();
    service = moduleRef.get<OpmlService>(OpmlService);

    const userRepository = moduleRef.get<Repository<UserEntity>>(getRepositoryToken(UserEntity));

    user = userRepository.create({
      id: 1,
      username: 'testuser',
      passwordHash: 'testpasswordhash',
    });
    await userRepository.save(user);
  });

  it('should be defined', () => {
    expect(service).toBeDefined();
  });

  it('should import feeds from OPML', async () => {
    const readable = fs.createReadStream(path.join(__dirname, 'feeds.opml'));
    await service.importFeeds(user, readable);

    const categoryRepository = moduleRef.get<Repository<CategoryEntity>>(getRepositoryToken(CategoryEntity));
    await expect(categoryRepository.find({ relations: { feeds: true } })).resolves.toMatchObject([
      {
        userId: user.id,
        name: 'Category 1',
        feeds: [
          { userId: user.id, title: 'Feed 1' },
          { userId: user.id, title: 'Feed 2' },
        ],
      },
      { userId: user.id, name: 'Category 2', feeds: [{ userId: user.id, title: 'Feed 3' }] },
    ]);
  });

  it('should import feeds without category from OPML', async () => {
    const readable = fs.createReadStream(path.join(__dirname, 'feeds-without-category.opml'));
    await service.importFeeds(user, readable);

    const categoryRepository = moduleRef.get<Repository<CategoryEntity>>(getRepositoryToken(CategoryEntity));
    await expect(categoryRepository.find({ relations: { feeds: true } })).resolves.toMatchObject([
      {
        userId: user.id,
        name: DEFAULT_CATEGORY_NAME,
        feeds: [
          { userId: user.id, title: 'Feed 1' },
          { userId: user.id, title: 'Feed 3' },
          { userId: user.id, title: 'Feed 5' },
        ],
      },
      {
        userId: user.id,
        name: 'Category 1',
        feeds: [{ userId: user.id, title: 'Feed 2' }],
      },
      {
        userId: user.id,
        name: 'Category 2',
        feeds: [{ userId: user.id, title: 'Feed 4' }],
      },
    ]);
  });
});
