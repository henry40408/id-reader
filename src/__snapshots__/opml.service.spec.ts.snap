// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OPML service should export feeds to OPML 1`] = `
"<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<opml version="2.0">
  <head>
    <title>Exported Feeds (2025-01-01T00:00:00.000Z)</title>
  </head>
  <body>
    <outline text="Category 1">
      <outline text="Feed 1" xmlUrl="http://example1.invalid/feed" htmlUrl="http://example1.invalid"/>
      <outline text="Feed 2" xmlUrl="http://example2.invalid/feed" htmlUrl="http://example2.invalid"/>
    </outline>
    <outline text="Category 2">
      <outline text="Feed 3" xmlUrl="http://example3.invalid/feed" htmlUrl="http://example3.invalid"/>
    </outline>
  </body>
</opml>"
`;
